Add kubernetes-1.28 package and variants

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Resolves https://github.com/bottlerocket-os/bottlerocket/issues/3274

**Description of changes:**
This adds the package for K8s 1.28 for building kubelet and new 1.28 Bottlerocket variants.

The new variants will use 6.1 kernel, include secure boot, and use XFS for the data volume filesystem.

**Testing done:**
- [x] aarch64 IPv4 conformance tests
- [x] aarch64 IPv6 conformance tests
- [x] aarch64 NVIDIA conformance tests
- [x] aarch64 NVIDIA smoke tests
- [x] x86_64 IPv4 conformance tests
- [x] x86_64 IPv6 conformance tests
- [x] x86_64 NVIDIA conformance tests
- [x] x86_64 NVIDIA smoke tests
```
$ testsys status 
 NAME                                                     TYPE   STATE     PASSED   SKIPPED   FAILED 
 beta-aarch64-aws-k8s-128-conformance-test                Test       passed      382      7007      2      
 beta-aarch64-aws-k8s-128-conformance-test-retry-1        Test       passed      4        7387      0      
 beta-aarch64-aws-k8s-128-instances                       Resource   completed                             
 beta-aarch64-aws-k8s-128-ipv6-conformance-test           Test       passed      381      7007      3      
 beta-aarch64-aws-k8s-128-ipv6-conformance-test-retry-1   Test       passed      4        7387      0      
 beta-aarch64-aws-k8s-128-ipv6-instances                  Resource   completed                             
 beta-aarch64-aws-k8s-128-nvidia-ipv6-conformance-test    Test       passed      384      7007      0      
 beta-aarch64-aws-k8s-128-nvidia-ipv6-instances           Resource   completed                             
 beta-aarch64-aws-k8s-128-nvidia-ipv6-nvidia-smoke-test   Test       passed      11       0         0      
 beta-x86-64-aws-k8s-128-conformance-test                 Test       passed      384      7007      0      
 beta-x86-64-aws-k8s-128-instances                        Resource   completed                             
 beta-x86-64-aws-k8s-128-ipv6-conformance-test            Test       passed      384      7007      0      
 beta-x86-64-aws-k8s-128-ipv6-instances                   Resource   completed                             
 beta-x86-64-aws-k8s-128-nvidia-ipv6-conformance-test     Test       passed      384      7007      0      
 beta-x86-64-aws-k8s-128-nvidia-ipv6-instances            Resource   completed                             
 beta-x86-64-aws-k8s-128-nvidia-ipv6-nvidia-smoke-test    Test       passed      11       0         0
```
- [x] `metal-k8s-1.28`
```bash
$ kubectl --kubeconfig br-128-eks-a-cluster.kubeconfig get nodes -o wide
NAME                                 STATUS   ROLES           AGE    VERSION               INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                                  KERNEL-VERSION   CONTAINER-RUNTIME
br-128-dfw9b                         Ready    control-plane   124m   v1.28.0-eks-bb809b9   10.80.50.26   <none>        Bottlerocket OS 1.15.0 (metal-k8s-1.28)   6.1.41           containerd://1.6.20+bottlerocket
br-128-md-0-64954d44f5xczblw-59hw7   Ready    <none>          109m   v1.28.0-eks-bb809b9   10.80.50.32   <none>        Bottlerocket OS 1.15.0 (metal-k8s-1.28)   6.1.41           containerd://1.6.20+bottlerocket
br-128-md-0-64954d44f5xczblw-nw8n6   Ready    <none>          109m   v1.28.0-eks-bb809b9   10.80.50.30   <none>        Bottlerocket OS 1.15.0 (metal-k8s-1.28)   6.1.41           containerd://1.6.20+bottlerocket

$ sonobuoy --kubeconfig br-128-eks-a-cluster.kubeconfig status 
   PLUGIN     STATUS   RESULT   COUNT                                PROGRESS
      e2e   complete   passed       1   Passed:380, Failed:  0, Remaining:  0

Sonobuoy has completed. Use `sonobuoy retrieve` to get results.

```
- [x] `vmware-k8s-1.28`
```
$ kubectl --kubeconfig br-eksa-128-eks-a-cluster.kubeconfig get nodes
NAME                                      STATUS   ROLES           AGE   VERSION
br-eksa-128-cgn4r                         Ready    control-plane   15m   v1.28.0-eks-bb809b9
br-eksa-128-j7vzq                         Ready    control-plane   14m   v1.28.0-eks-bb809b9
br-eksa-128-md-0-6bb48f8cb5x4h8gw-5z2k2   Ready    <none>          14m   v1.28.0-eks-bb809b9
br-eksa-128-md-0-6bb48f8cb5x4h8gw-z8lvw   Ready    <none>          14m   v1.28.0-eks-bb809b9
...
Server Version: version.Info{Major:"1", Minor:"28", GitVersion:"v1.28.0-eks-bb809b9", GitCommit:"bb809b97e4d30ba0c4f4e7905e559cb20e0ada33", GitTreeState:"archive", BuildDate:"2023-08-15T10:15:49Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}
$ sonobuoy --kubeconfig br-eksa-128-eks-a-cluster.kubeconfig run --plugin=e2e --mode=certified-conformance --wait
19:13:14       e2e   global   complete   passed   Passed:380, Failed:  0, Remaining:  0
19:13:14 Sonobuoy has completed. Use `sonobuoy retrieve` to get results.
```
- [x] EBS driver testing
```
$ kubectl --kubeconfig beta-aarch64-aws-k8s-128-ipv6.kubeconfig get pods -n kube-system -l app.kubernetes.io/name=aws-ebs-csi-driver
NAME                                  READY   STATUS    RESTARTS   AGE
ebs-csi-controller-55d4d546d8-9c6qz   5/5     Running   0          20m
ebs-csi-controller-55d4d546d8-xxzr8   5/5     Running   0          20m
ebs-csi-node-5l6b9                    3/3     Running   0          20m
ebs-csi-node-tkcgc                    3/3     Running   0          20m
```
- [x] ECR Credential Provider testing (thanks @stmcginnis )

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
packages: add kubernetes-1.28

This adds the K8s 1.28 package.

etungsten committed:
sources, variants: add aws-k8s-1.28 variant

etungsten committed:
sources, variants: add aws-k8s-1.28-nvidia variant

etungsten committed:
sources, variants: add vmware-k8s-1.28 variant

etungsten committed:
sources, variants: add metal-k8s-1.28 variant

stmcginnis commented on packages/kubernetes-1.28/dockershim-symlink.conf:
```diff
@@ -0,0 +1,2 @@
+[Service]
+ExecStartPre=/bin/ln -sf /run/containerd/containerd.sock /run/dockershim.sock
```
This may not be necessary anymore. I started to look in to it when the last variants were added, but will have to look in my notes to see what the situation was then.

etungsten replied:
The symlink is there to allow container workloads that still assume the container runtime socket to be `dockershim.sock` to continue to work. I'm not opposed to removing this, but maybe we should start with a deprecation notice before complete removal? not sure.

stmcginnis commented on variants/Cargo.toml:
```diff
@@ -27,6 +29,7 @@ members = [
     "vmware-k8s-1.25",
     "vmware-k8s-1.26",
     "vmware-k8s-1.27",
+    "vmware-k8s-1.28",
```
Missing `metal-k8s-1.28`?

stmcginnis commented:
Added `aws-k8s-1.28` to a 1.27 EKS cluster:

```
$ k get nodes
NAME                                           STATUS   ROLES    AGE    VERSION
ip-192-168-4-253.us-east-2.compute.internal    Ready    <none>   4m7s   v1.28.0-rc.0-eks-506ffcd
ip-192-168-41-249.us-east-2.compute.internal   Ready    <none>   4m6s   v1.28.0-rc.0-eks-506ffcd
```

Attempted to deploy pod with image hosted in private registry of a different account:

```
$ k get pod
NAME     READY   STATUS         RESTARTS   AGE
brtest   0/1     ErrImagePull   0          4s
```

Applied `credential-provider` settings:

```
apiclient apply <<EOF
[settings.kubernetes.credential-providers.ecr-credential-provider]
enabled = true
cache-duration = "30m"
image-patterns = [
  "*.dkr.ecr.us-east-2.amazonaws.com",
  "*.dkr.ecr.us-west-2.amazonaws.com"
]

[settings.aws]
profile = "ecr"
config = “W3...o=”
EOF
```

Checked `kubelet` status and logs and verified image able to be pulled and run:

```
$ k get pod
NAME     READY   STATUS    RESTARTS   AGE
brtest   1/1     Running   0          23m
```

Ran `sonobuoy run --mode=certified-conformance --wait` and verified tests passed:

```
15:34:34             e2e                                         global   complete   passed   Passed:378, Failed:  0, Remaining:  0
15:34:34    systemd-logs    ip-192-168-4-253.us-east-2.compute.internal   complete   passed
```

jpculp commented:
We'll probably want to add `seccomp-default` to `kubelet-config` if https://github.com/bottlerocket-os/bottlerocket/pull/3334 merges first.

bcressey commented:
Will need to rebase on the changes in #3259 also which changes the `76-oci-defaults-capabilities.toml` symlink.

stmcginnis commented on packages/kubernetes-1.28/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+# RPM Requires
+[dependencies]
+aws-signing-helper = { path = "../aws-signing-helper" }
+ecr-credential-provider-1_27 = { path = "../ecr-credential-provider-1.27" }
```
The 1.28 provider is tagged. I don't see anything changing with that release for the credential provider, but it would be good to pick that up.

etungsten replied:
I'm confused. Are we suppose to include the provider in lockstep with the K8s version? Are all K8s variants suppose to use the latest provider version?

stmcginnis replied:
Officially, yes. The ECR credential provider version is supposed to match the K8s version.

Unofficially, very little changes with the interface to this, so up until the 1.27 variant we didn't have any problem using the same release for each variant. With 1.27 the version of the request/response changed, so we had to update to the newer provider.

Nothing actually changes between 1.27 and 1.28, so we _could_ still use the 1.27 provider with the newer variant. But the official support statement from the provider team is that they should match.

I'm really fine either way, so whichever direction you want to go is actually fine with me. It would be nice not to have to add yet another package that - when it comes down to it - is just duplication. I think we just need to decide if we are going to do this 1 for 1, or just be prepared that when adding a new variant we need to double check and make sure we don't also need to add a new provider.

etungsten replied:
If that's the case I'd rather not add another package and treat the ecr credential provider updates as a separate task.

stmcginnis reviewed (approved):
Everything looks good. Not much has changed since my last test, but built and added node to a 1.27 cluster.

```
NAME                                           STATUS   ROLES    AGE   VERSION
ip-192-168-66-218.us-east-2.compute.internal   Ready    <none>   63s   v1.28.0-eks-bb809b9
```

```
# sonobuoy run --mode=quick --wait
...
19:42:46             e2e                                         global   complete   passed   Passed:  1, Failed:  0, Remaining:  0
19:42:46    systemd-logs   ip-192-168-66-218.us-east-2.compute.internal   complete   passed
```

etungsten committed:
variants: update README with new control plane and node version skew

Starting with K8s 1.28, the supported skew between the node and control
plane componenets expands by one minor version from n-2 to n-3.

See https://kubernetes.io/blog/2023/08/15/kubernetes-v1-28-release/#changes-to-supported-skew-between-control-plane-and-node-versions

etungsten commented:
EKS-D is planning to do another release for 1.28.1 sometime tomorrow(?) but there are no kubelet changes and the only change is a windows CVE fix. I'll grab the update if it's available and if this doesn't merge by tomorrow.

etungsten commented:
If we update to the next EKS-D release, we can drop the `0001-Make-gomaxprocs-install-optional-limit-to-tests.patch` patch since it's included in 1.28.1: https://github.com/kubernetes/kubernetes/commit/cce3f6ffa796ea416021b0d62567a3f52b979567

bcressey reviewed (commented):
It looks like the `RotateKubeletServerCertificate` feature gate was dropped - is this on by default in 1.28?

stmcginnis commented:
> It looks like the `RotateKubeletServerCertificate` feature gate was dropped - is this on by default in 1.28?

Yeah, this was actually [enabled by default since K8s 1.12](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) (!), but I think it's good to only make the change for the new variant.

bcressey commented:
> Yeah, this was actually [enabled by default since K8s 1.12](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) (!), but I think it's good to only make the change for the new variant.

Is there any output that can be used to confirm that the feature gate is still enabled?

etungsten commented:
> > Yeah, this was actually [enabled by default since K8s 1.12](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) (!), but I think it's good to only make the change for the new variant.
> 
> Is there any output that can be used to confirm that the feature gate is still enabled?

There apparently is no way to query kubelet at runtime to see what feature gates are enabled/disabled.
We can however check the feature gate defaults in k8s source: https://github.com/kubernetes/kubernetes/blob/v1.28.0/pkg/features/kube_features.go#L1136
